File: cpp/336.cpp
class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        int size = words.size();
        vector<vector<int>> ret;
        unordered_map<string, int> dict;
        for (int i = 0; i < size; ++i) {
            string key = words[i];
            reverse(key.begin(), key.end());
            dict[key] = i;
        }
        // edge case: if empty string "" exists, find all palindromes to become pairs ("", self)
        if (dict.find("") != dict.end()) {
            for (int i = 0; i < size; ++i) {
                if (i == dict[""])
                    continue;
                if (isPalindrome(words[i]))
                    ret.push_back({dict[""], i});
            }
        }
        for (int i = 0; i < size; ++i) {
            int wl = words[i].size();
            for (int j = 0; j < wl; ++j) {
                string left = words[i].substr(0, j);
                string right = words[i].substr(j, wl - j);
                if (dict.find(left) != dict.end() && isPalindrome(right) && dict[left] != i)
                    ret.push_back({i, dict[left]});
                if (dict.find(right) != dict.end() && isPalindrome(left) && dict[right] != i)
                    ret.push_back({dict[right], i});
            }
        }
        return ret;
    }

    bool isPalindrome(const string s) {
        int size = s.size();
        for (int i = 0; i < size / 2; ++i)
            if (s[i] != s[size-i-1])
                return false;
        return true;
    }
};

File: go/336.go
func palindromePairs(words []string) [][]int {
    ret, dict := [][]int{}, map[string]int{}
    for i, word := range words {
        dict[reverseString(word)] = i
    }
    if index, ok := dict[""]; ok {
        for i, word := range words {
            if i == index {
                continue
            }
            if isPalindrome(word) {
                ret = append(ret, []int{index, i})
            }
        }
    }
    for i, word := range words {
        wl := len(word)
        for j := 0; j < wl; j++ {
            left, right := word[:j], word[j:]
            if index, ok := dict[left]; ok && isPalindrome(right) && index != i {
                ret = append(ret, []int{i, index})
            }
            if index, ok := dict[right]; ok && isPalindrome(left) && index != i {
                ret = append(ret, []int{index, i})
            }
        }
    }
    return ret
}

func reverseString(str string) string {
    s, size := []byte(str), len(str)
    for i := 0; i < size/2; i++ {
        s[i], s[size-i-1] = s[size-i-1], s[i]
    }
    return string(s)
}

func isPalindrome(str string) bool {
    size := len(str)
    for i := 0; i < size/2; i++ {
        if str[i] != str[size-i-1] {
            return false
        }
    }
    return true
}
